BIM: Introduce 1-axis mode for ortho array

tetektoza opened the pull request:
As the title says - the 1 axis mode allows to switch between all of the axises mode and allows to modify only 1 axis at the time that user can select with the checkbox.

@kaiwas is this what you were expecting from this feature?

Demo:

https://github.com/user-attachments/assets/e857d9e2-32fc-43d6-9751-57c24f7031e0


Resolves: https://github.com/FreeCAD/FreeCAD/issues/16865

tetektoza commented:
I've labeled it as BIM, because Array is not by default available here I think, and the issue was labeled under Draft/BIM too.

I can also reduce amount of intervals here, like just "X" instead of XYZ for X axis. Open for suggestions before I remove my prints from this PR

furgo16 commented:
The feature is great (thanks!), and I think it would greatly benefit from @FreeCAD/design-working-group input.

Some thoughts:

1. It noticeably simplifies array creation (great). While I don't have any data to back this up, if I were to make an educated guess, most of the arrays created for BIM purposes are along one axis only.
2. In general though, it feels like an iteration over a UX that was already complicated to understand in the first place. I wonder if we can do better than patching the current behaviour.
3. The "1-Axis mode" button name seems too technical. "One-dimensional array"? What do other CAD/BIM software packages use for this term?
4. Whichever implementation of the "1-Axis mode" widget (a button in the initial video), it needs a tooltip.
5. Yes, I think if we simplify the axes, we should simplify the intervals as well. So if I want to create an array of beams along the X axis, I will only want to specify the x interval/spacing, and the rest should be hidden (but "unhideable" if someone needs them). A "Uniform spacing" checkbox checked by default, with an input box next to it could be an option. When unchecked, it would disable the input box, and show three independent x,y,z input boxes.

tetektoza commented:
@yorikvanhavre you can give it a test but don't merge yet Yorik, I wanna apply some of guys' points, but I was screwing around with Coin last days and didn't have time to get back to this one. Will try to do something about it today

yorikvanhavre commented:
Ok, no hurry ;)
I'll mark this as Draft so there is no risk to have it merged accidentally. Just press the "ready for review" button when you're ready!

tetektoza commented:
Added stuff:
- caching of variables (now we cache Number of elements, intervals, mode that Orthogonal Array is currently in (Linear/Ortho) - check demo)
- changed name from 1-Axis mode to Linear Mode
- now we only have one axis for intervals (you select Y axis -> you get only Y intervals
- made the checkboxes more exclusive, as there was a case where user could deselect the checkbox at all, now they can't do it, at least one has to be selected

@Roy-043 you're also welcome to review it a bit so maybe this PR will be less invasive for your changes, btw thx for putting your task on hold :D

Demo attached:

https://github.com/user-attachments/assets/e790a3c1-0e66-4518-b815-a0b9890d8b4d

marcuspollio commented:
@tetektoza Nice improvement!

Could you please rename your PR title and amend the commits names to indicate the changes are in the `Draft` module, not `BIM`.

Now I'm wondering from a UX perspective: this PR introduced the `Mode` concept for arrays.
We already have the `Array Type` concept (`ortho`, `polar` and `circular`) that can be changed later on via the Property View panel. This new Linear Mode is not really a new Type, as it only changes how the Orthogonal Array options (axes intervals) are shown to users. Is there nonetheless a way to somehow harmonize UX without introducing this second mode concept?
Maybe @Roy-043 has some ideas? Or maybe it's just me and it's not a big deal? =D

tetektoza commented:
Changed a couple of things according to review:
* linear mode is now being used as default during first startup (it
  wasn't before)
* applied Roy's comments about coding style, etc.
* grouped everything into separate QGroupBox which is dedicated for
  Linear Mode and has it's own labels, although spinboxes are shared so
  during mode switch we reparent them now
* removed Orthogonal Array's icon
* in the QGroup applied naming suggested in the review, X Intervals ->
  interval, etc.
* changed to radio buttons since we want exclusivity in selection

Demo:

https://github.com/user-attachments/assets/1e202071-f689-411b-be9a-ba2aef2e1644

tetektoza commented on src/Mod/Draft/Resources/ui/TaskPanel_OrthoArray.ui:
```diff
@@ -46,6 +46,62 @@
+         <item row="1" column="0">
+          <layout class="QGridLayout" name="grid_axis_select">
+           <item row="0" column="0">
+            <widget class="QCheckBox" name="checkbox_x_axis">
```
Good point, sometimes I go with one solution and try to force it to behave one way instead of switching idea. Changed to radio buttons.

tetektoza commented on src/Mod/Draft/Resources/ui/TaskPanel_OrthoArray.ui:
```diff
@@ -46,6 +46,62 @@
       <item row="1" column="0">
+       <widget class="QGroupBox" name="group_axis_mode">
+        <property name="toolTip">
+         <string>Toggle between Orthogonal mode and Linear mode. In Linear mode, you can select which axes to use.</string>
```
Done.

tetektoza commented on src/Mod/Draft/draftutils/params.py:
```diff
@@ -506,6 +506,18 @@ def _get_param_dictionary():
+        "Z_interval": ("float", start_val),
+        "X_NumOfElements": ("int", 1),
+        "Y_NumOfElements": ("int", 1),
+        "Z_NumOfElements": ("int", 1)
```
Removed underscores, capitalized interval, added default number of X Y and Z elems to be 2, also made Linear Mode to be True by default as someone requested.

tetektoza commented on src/Mod/Draft/drafttaskpanels/task_orthoarray.py:
```diff
@@ -357,6 +420,10 @@ def print_messages(self):
         _msg(translate("draft","Object:") + " {}".format(sel_obj.Label))
+
+        if self.linear_mode and self.has_axis_ui:
+            _msg(translate("draft","Using Linear mode"))
```
OK, removed.

tetektoza commented on src/Mod/Draft/drafttaskpanels/task_orthoarray.py:
```diff
@@ -234,6 +284,19 @@ def create_object(self):
+                self.v_y = App.Vector(0, 0, 0)
+            if not self.form.checkbox_z_axis.isChecked():
+                self.n_z = 1
+                self.v_z = App.Vector(0, 0, 0)
```
OK, changed.

tetektoza commented on src/Mod/Draft/drafttaskpanels/task_orthoarray.py:
```diff
@@ -143,6 +144,26 @@ def __init__(self):
+        self.has_axis_ui = hasattr(self.form, 'button_linear_mode') and \
+                          hasattr(self.form, 'checkbox_x_axis') and \
+                          hasattr(self.form, 'checkbox_y_axis') and \
+                          hasattr(self.form, 'checkbox_z_axis')
```
Removed this completely as I've took your suggestion that maybe it's pointless to check if we have those UI elements.

tetektoza commented on src/Mod/Draft/drafttaskpanels/task_orthoarray.py:
```diff
@@ -143,6 +144,26 @@ def __init__(self):
+        self.active_axis = params.get_param("AxisSelected", "Mod/Draft/OrthoArrayLinearMode")
+
+        # Check if the UI elements exist before trying to access them
+        self.has_axis_ui = hasattr(self.form, 'button_linear_mode') and \
```
Removed this completely as I've took your suggestion that maybe it's pointless to check if we have those UI elements.

tetektoza commented on src/Mod/Draft/Resources/ui/TaskPanel_OrthoArray.ui:
```diff
@@ -46,6 +46,62 @@
+         <string>Toggle between Orthogonal mode and Linear mode. In Linear mode, you can select which axes to use.</string>
+        </property>
+        <property name="title">
+         <string>Axis Mode</string>
```
Changed to sentence case, you can check rest.

furgo16 commented:
This looks good to go to me UX-wise. Thanks for the great work!

Roy-043 commented:
Thanks also from my side. This can be merged. Since this has been assigned to Yorik I'll let him do the honors.

yorikvanhavre commented:
I took the liberty to squash the commits @tetektoza , I hope thus is fine,  tell me if you prefer not to, next time!

PaddleStroke commented:
This UI with a button is very exotic and inconsistent with the rest of freecad.
The way it is usually done is to have a Mode combobox at the beginning.

tetektoza commented:
>On top of that the button is in a groupbox. Why?

@PaddleStroke Why? Because no one said anything about it, just after merging the PR you did. There was a discussion regarding this as well on Discord and this topic, that's why I also provide demo to those PRs to be sure we agree on something. The button is in the group box because the radiobuttons are with it, why I've chosen this approach I have no bigger explanation, it looks odd when we hide those radiobuttons, but I can change it, on top of the button to be a combobox or w/e else if everyone agrees.

PaddleStroke commented:
> Pay attention to the "OK/Apply/Cancel/Help/Close" buttons. They are all different. Doesn't that bother you? But for some reason the switch bothers you.

That makes no sense. It's not because I comment on this that other things do not bother me.

The mode combobox is this : 
![image](https://github.com/user-attachments/assets/1e570622-932a-4a2f-980e-4e8238a17b5b)

![image](https://github.com/user-attachments/assets/733ddec0-913b-4468-9408-f5ca8c120948)

![image](https://github.com/user-attachments/assets/0992daac-8d21-4d87-865d-8f034aa9d3a9)

![image](https://github.com/user-attachments/assets/bb0161c4-f710-4017-afd7-8ea82aab5c13)

furgo16 commented:
I would suggest to continue the discussion in a separate issue or PR. There was a rich and civil discussion, we reached agreement and @tetektoza kindly factored in all feedback. The beauty of open source is that we can always make iterative improvements.
